### PR TITLE
TMS5220 changes, bugfixes and updates

### PR DESCRIPTION
**Pull request by Lord-Nightmare**
PR as requested, with changes (roughly) split.

### CODE REVIEW

**Comment by Lord-Nightmare** [comment]
With that last fix (1186c07, of two lines missed during cherry-picking lines for the various commits), everything listed here was tested and works: 720, monymony, jackrabt, starwars, jedi, gauntlet, paperboy, roadrunn, eprom, dotrone, ti99_4a(in both vsm and speak external mode), apple2 (echoii)

**Comment by Tafoid** [comment]
I took some time @Lord-Nightmare 's commit and made a 32-bit version covering all drivers using the tms5220 source file.  I was also provided binaries for MAME and MESS 64-bit targets.  In comparison to current wavdata, only a few machines showed any type of data change (gaunt2, indytemp, roadblst, roadrunn).  When I manually tested each one, I did not detect any type of breakage or shortcoming with the current presentation.  With my testing and followup, I endorse that this PR does not appear to break functionality of the TMS5220 in any way.

File: src/devices/sound/tms5220.cpp [hunk 1]
```diff
@@ -815,54 +839,14 @@ int tms5220_device::status_read()
...
-int tms5220_device::cycles_to_ready()
-{
-	int answer;
-
-
-	if (ready_read())
-		answer = 0;
+	// if m_true_timing is NOT set (we're in 'hacky instant write mode'), the m_timer_io_ready doesn't run and will never de-assert m_io_ready if the fifo is full, so we need to explicitly check for fifo full here and return the proper value.
+	if (!m_true_timing)
+		return ((m_fifo_count < FIFO_SIZE)||(!m_DDIS)) && m_io_ready; // SEVERE CAVEAT: this makes the assumption that the ready_read was after wsq was 'virtually asserted', so if the fifo has no room in it ready will always return inactive, even if no write happened (i.e., after a read command when the fifo was exactly filled but no write attempted to overfill it) which is inaccurate to hardware and may cause issues! you have been warned!
```
Comment (cuavas) [inline src/devices/sound/tms5220.cpp]: Please wrap comments somewhere around column 72 - it makes them a lot more readable.

File: src/devices/sound/tms5220.cpp [hunk 2]
```diff
@@ -1972,13 +1962,15 @@ void tms5220_device::write_data(uint8_t data)
...
 
 	if (!m_true_timing)
 	{
-		/* bring up to date first */
-		m_stream->update();
-		return status_read();
+		// prevent debugger from changing the internal state
+		if (machine().side_effects_disabled())
+			return status_read(false);
+		return status_read(true);
```
Comment (cuavas) [inline src/devices/sound/tms5220.cpp]: You could just do `return status_read(!machine().side_effects_disabled());` here.

**Comment by Lord-Nightmare** [comment]
The thing which broke jedi was actually a bug in the original (reverted) commit, where I'd forgotten to re-add the hack with the old non-wsq/rsq (i.e. not m_true_timing) mode. The hack checks the fifo fullness and returning the not-ready state on a direct call to ready_read. AJR's updating jedi.cpp bypassed the bug by switching it to the new m_true_timing interface, and I re-added the hack in 7ed1eb1 (with that long comment which I should break into a few lines) anyway.

mhavocrv I'm actually working on getting the major havoc pcb traced, since the schematics omit the tms5220 and its related circuitry entirely. (I should probably add some comments to mhavoc.cpp about how the chip is miswired on the mhavoc pcb as shipped and requires several greenwire fixes to even work.
Notably, VSS and VDD are swapped backwards, and the speaker is tapped off the wrong pin as well.
It is notable that the jedi schematics show VSS/VDD backwards as well, though in that case the final pcb had the swap fixed, I think.

**Comment by Lord-Nightmare** [comment]
turbosub I'll poke phil_b about, since based on comments, it sounds like the real PCB uses a state machine PROM to handle asserting /WS, driving the buffer chip to the tms5220, etc, and best to properly emulate that than relying on hacks.

File: src/devices/sound/tms5220.cpp (hunk 1 above)
Comment (Lord-Nightmare) [reply src/devices/sound/tms5220.cpp]: will do. I'll also make "FIFO" consistently capitalized in comments and logs since its horribly inconsistent.

**Pull request merged**